### PR TITLE
feat(gpu): add support to larger polynomials on multi-bit PBS

### PR DESCRIPTION
**Pull request by pdroalves**
closes: https://github.com/zama-ai/tfhe-rs-internal/issues/455

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by pdroalves** [comment]
@slab-ci gpu-test

**Comment by pdroalves** [comment]
@slab-ci gpu_test

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_multibit.cuh [hunk 1]
```diff
@@ -245,38 +297,68 @@ __host__ void host_cg_multi_bit_programmable_bootstrap(
...
-        stream->stream));
+    // Compute a keybundle
+    execute_compute_keybundle<Torus, params>(
+        stream, lwe_array_in, lwe_input_indexes, bootstrapping_key, buffer,
+        num_samples, lwe_dimension, glwe_dimension, polynomial_size,
+        grouping_factor, base_log, level_count, max_shared_memory,
+        lwe_chunk_size, lwe_offset);
+
+    // Accumulate
+    execute_accumulate_step<Torus, params>(
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_multibit.cuh:357]: I think I would have named it `execute_external_product_loop` maybe.

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cuh [hunk 2]
```diff
@@ -152,11 +162,23 @@ __global__ void device_multi_bit_programmable_bootstrap_accumulate_step_one(
...
+
   Torus *accumulator = (Torus *)selected_memory;
   double2 *accumulator_fft =
       (double2 *)accumulator +
       (ptrdiff_t)(sizeof(Torus) * polynomial_size / sizeof(double2));
 
+  if constexpr (SMD == PARTIALSM)
+    accumulator_fft = (double2 *)sharedmem;
+
+  //
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cuh:181]: Unnecessary line :slightly_smiling_face:

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cuh [hunk 3]
```diff
@@ -219,22 +241,33 @@ __global__ void device_multi_bit_programmable_bootstrap_accumulate_step_one(
...
+    selected_memory = sharedmem;
+  } else {
+    int block_index = blockIdx.x + blockIdx.y * gridDim.x +
+                      blockIdx.z * gridDim.x * gridDim.y;
+    selected_memory = &device_mem[block_index * device_memory_size_per_block];
+  }
+
   double2 *accumulator_fft = (double2 *)selected_memory;
 
+  //
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cuh:270]: Same here

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cuh [hunk 4]
```diff
@@ -390,84 +478,187 @@ __host__ void scratch_multi_bit_programmable_bootstrap(
...
+  uint32_t keybundle_size_per_input =
+      lwe_chunk_size * level_count * (glwe_dimension + 1) *
+      (glwe_dimension + 1) * (polynomial_size / 2);
 
-  //
   uint64_t full_sm_keybundle =
       get_buffer_size_full_sm_multibit_programmable_bootstrap_keybundle<Torus>(
           polynomial_size);
+
+  //
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cuh:501]: Same here

**Review by agnesLeroy — COMMENTED** [review]
Hey @pdroalves! Looks really good, thank you very much! I just have a comment about one name choice, otherwise all good :slightly_smiling_face:

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_multibit.cuh (hunk 1 above)
Comment (pdroalves) [reply backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_multibit.cuh:357]: Perfect!

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cuh (hunk 2 above)
Comment (pdroalves) [reply backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cuh:181]: The `//`?

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_multibit.cuh [hunk 5]
```diff
@@ -195,41 +248,40 @@ __host__ void scratch_cg_multi_bit_programmable_bootstrap(
...
+__host__ void execute_external_product_loop(
+    cuda_stream_t *stream, Torus *lut_vector, Torus *lut_vector_indexes,
+    Torus *lwe_array_in, Torus *lwe_input_indexes, Torus *lwe_array_out,
+    Torus *lwe_output_indexes, pbs_buffer<Torus, MULTI_BIT> *buffer,
+    uint32_t num_samples, uint32_t lwe_dimension, uint32_t glwe_dimension,
+    uint32_t polynomial_size, uint32_t grouping_factor, uint32_t base_log,
+    uint32_t level_count, uint32_t lwe_chunk_size, uint32_t max_shared_memory,
+    int lwe_offset) {
 
   //
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_multibit.cuh:261]: Another one here

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_multibit.cuh [hunk 6]
```diff
@@ -195,41 +248,40 @@ __host__ void scratch_cg_multi_bit_programmable_bootstrap(
...
+  uint64_t no_dm = 0;
 
   uint32_t keybundle_size_per_input =
       lwe_chunk_size * level_count * (glwe_dimension + 1) *
       (glwe_dimension + 1) * (polynomial_size / 2);
 
+  uint32_t chunk_size =
+      std::min(lwe_chunk_size, (lwe_dimension / grouping_factor) - lwe_offset);
+
+  //
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_multibit.cuh:277]: Here as well

File: backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_multibit.cuh [hunk 7]
```diff
@@ -195,41 +248,40 @@ __host__ void scratch_cg_multi_bit_programmable_bootstrap(
...
+  auto d_mem = buffer->d_mem_acc_cg;
+  auto keybundle_fft = buffer->keybundle_fft;
+  auto global_accumulator = buffer->global_accumulator;
+  auto buffer_fft = buffer->global_accumulator_fft;
+
   //
```
Comment (agnesLeroy) [inline backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_cg_multibit.cuh:283]: Here as well

**Review by agnesLeroy — APPROVED** [review]

**Pull request merged**